---
title: "Molecular subtyping for ovarian cancer"
author: "Gregory M Chen"
date: "6/9/2017"
output:
  html_document: default
  pdf_document: default
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
```


## Introduction

`consensusOV` is a package for molecular subtyping for ovarian cancer. It is intended for whole-transcriptome gene expression datasets from patients with high-grade serous ovarian carcinoma. This package includes implementations of four previously published subtype classifiers [@konecny2014prognostic; @verhaak2013prognostically; @bentink2012angiogenic; @helland2011deregulation] and a consensus random forest classifier.

The `get.subtypes()` function is a wrapper for the other package subtyping functions  `get.consensus.subtypes()`, `get.konecny.subtypes()`, `get.verhaak.subtypes()`,
`get.bentink.subtypes()`, `get.helland.subtypes()`. It can take as input either a matrix of gene expression values and a vector of Entrez IDs; or a BioBase::ExpressionSet following the format of MetaGxOvarian [@gendoo2016metagxdata]. If `expression.dataset` is a matrix, it should be formatted with genes as rows and patients as columns; and `entrez.ids` should be a vector with length the same as `nrow(expression.dataset)`. The `method` argument specifies which of the five subtype classifiers to use.

## Load Data

The package contains a subset of `GSE14764` as sample data.

```{r load_data, results='hide'}
library(consensusOV)
library(Biobase)
library(genefu)
data(GSE14764.eset)
GSE14764.expression.matrix <- exprs(GSE14764.eset)
GSE14764.entrez.ids <- fData(GSE14764.eset)$EntrezGene.ID
```


```{r view_dimensions, results='asis'}
dim(GSE14764.eset)
```

## Subtyping

```{r subtyping, results="hide"}
bentink.subtypes <- get.subtypes(GSE14764.eset, method = "Bentink")
konecny.subtypes <- get.subtypes(GSE14764.eset, method = "Konecny")
helland.subtypes <- get.subtypes(GSE14764.eset, method = "Helland")
verhaak.subtypes <- get.subtypes(GSE14764.eset, method = "Verhaak")
consensus.subtypes <- get.subtypes(GSE14764.eset, method = "consensusOV")
```

```{r alternative_subtyping, eval=FALSE}
## Alternatively:
bentink.subtypes <- get.subtypes(GSE14764.expression.matrix, GSE14764.entrez.ids, method = "Bentink")
konecny.subtypes <- get.subtypes(GSE14764.expression.matrix, GSE14764.entrez.ids, method = "Konecny")
helland.subtypes <- get.subtypes(GSE14764.expression.matrix, GSE14764.entrez.ids, method = "Helland")
verhaak.subtypes <- get.subtypes(GSE14764.expression.matrix, GSE14764.entrez.ids, method = "Verhaak")
consensus.subtypes <- get.subtypes(GSE14764.expression.matrix, GSE14764.entrez.ids, method = "consensusOV")
```

Each subtyping function outputs a list with two values. The first value is a factor of subtype labels. The second is an classifier-specific values. For the Konecny, Helland, Verhaak, and Consensus classifiers, this object is a dataframe with subtype specific scores. For the Bentink classifier, the object is the output of the `genefu` function call. 

Subtype classifiers can alternatively be called using inner functions.

```{r subtyping_ex_2, eval=FALSE}
bentink.subtypes <- get.bentink.subtypes(GSE14764.expression.matrix, GSE14764.entrez.ids)
```

## Subtype Scores

The Konecny, Helland, Verhaak, and Consensus classifiers produce real-valued subtype scores. We can use these in various ways - for example, here, we compute correlations between correspinding subtype scores.

We can compare the subtype scores between the Verhaak and Helland classifiers:

```{r verhaak_helland, fig.height = 6, fig.width = 6}
par(mfrow=c(2,2))
plot(verhaak.subtypes$gsva.out[,"IMR"], helland.subtypes$subtype.scores[,"C2"], xlab="IMR", ylab="C2")
plot(verhaak.subtypes$gsva.out[,"DIF"], helland.subtypes$subtype.scores[,"C4"], xlab="DIF", ylab="C4")
plot(verhaak.subtypes$gsva.out[,"PRO"], helland.subtypes$subtype.scores[,"C5"], xlab="PRO", ylab="C5")
plot(verhaak.subtypes$gsva.out[,"MES"], helland.subtypes$subtype.scores[,"C1"], xlab="MES", ylab="C1")
```

Corresponding Spearman correlation values are `r cor(verhaak.subtypes$gsva.out[,"IMR"], helland.subtypes$subtype.scores[,"C2"], method="spearman")`, `r cor(verhaak.subtypes$gsva.out[,"DIF"], helland.subtypes$subtype.scores[,"C4"])`, `r cor(verhaak.subtypes$gsva.out[,"PRO"], helland.subtypes$subtype.scores[,"C5"])`, `r cor(verhaak.subtypes$gsva.out[,"MES"], helland.subtypes$subtype.scores[,"C1"])`.


Likewise, we can compare the subtype scores between the Konecny and Helland classifier:

```{r konecny_helland, fig.height = 6, fig.width = 6}
par(mfrow=c(2,2))
plot(konecny.subtypes$spearman.cc.vals[,"C1_immL"], helland.subtypes$subtype.scores[,"C2"], xlab="C1_immL", ylab="C2")
plot(konecny.subtypes$spearman.cc.vals[,"C2_diffL"], helland.subtypes$subtype.scores[,"C4"], xlab="C2_diffL", ylab="C4")
plot(konecny.subtypes$spearman.cc.vals[,"C3_profL"], helland.subtypes$subtype.scores[,"C5"], xlab="C3_profL", ylab="C5")
plot(konecny.subtypes$spearman.cc.vals[,"C4_mescL"], helland.subtypes$subtype.scores[,"C1"], xlab="C4_mescL", ylab="C1")
```

Corresponding Spearman correlation values are `r cor(konecny.subtypes$spearman.cc.vals[,"C1_immL"], helland.subtypes$subtype.scores[,"C2"])`, `r cor(konecny.subtypes$spearman.cc.vals[,"C2_diffL"], helland.subtypes$subtype.scores[,"C4"])`, `r cor(konecny.subtypes$spearman.cc.vals[,"C3_profL"], helland.subtypes$subtype.scores[,"C5"])`, `r cor(konecny.subtypes$spearman.cc.vals[,"C4_mescL"], helland.subtypes$subtype.scores[,"C1"])`

## References